urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/agent/evolution/NeuralNetworkDrivenAgent.java

```diff
@@ -0,0 +1,255 @@
+					nearestFoodDist = currFoodDist;
+				}
+			}
+		}
+
+		// Find nearest agent
+		Agent nearestAgent = null;
+		double nearestAgentDist = maxAgentsDistance;
+
+		for (Agent currAgent : environment.filter(Agent.class)) {
+			// agent can see only ahead
+			if ((this != currAgent) && (this.inSight(currAgent))) {
+				double currAgentDist = this.distanceTo(currAgent);
+				if (currAgentDist <= nearestAgentDist) {
+					nearestAgent = currAgent;
+					nearestAgentDist = currAgentDist;
+				}
+			}
+		}
+
+		List<Double> nnInputs = new LinkedList<Double>();
+
+		double rx = this.getRx();
+		double ry = this.getRy();
+
+		double x = this.getX();
+		double y = this.getY();
+
+		if (nearestFood != null) {
+			double foodDirectionVectorX = nearestFood.getX() - x;
+			double foodDirectionVectorY = nearestFood.getY() - y;
+
+			// left/right cos
+			double foodDirectionCosTeta =
+					Math.signum(this.pseudoScalarProduct(rx, ry, foodDirectionVectorX, foodDirectionVectorY))
+							* this.cosTeta(rx, ry, foodDirectionVectorX, foodDirectionVectorY);
+
+			nnInputs.add(FOOD);
+			nnInputs.add(nearestFoodDist);
+			nnInputs.add(foodDirectionCosTeta);
+
+		} else {
+			nnInputs.add(EMPTY);
+			nnInputs.add(0.0);
+			nnInputs.add(0.0);
+		}
+
+		if (nearestAgent != null) {
+			double agentDirectionVectorX = nearestAgent.getX() - x;
+			double agentDirectionVectorY = nearestAgent.getY() - y;
+
+			// left/right cos
+			double agentDirectionCosTeta =
+					Math.signum(this.pseudoScalarProduct(rx, ry, agentDirectionVectorX, agentDirectionVectorY))
+							* this.cosTeta(rx, ry, agentDirectionVectorX, agentDirectionVectorY);
+
+			nnInputs.add(AGENT);
+			nnInputs.add(nearestAgentDist);
+			nnInputs.add(agentDirectionCosTeta);
+
+		} else {
+			nnInputs.add(EMPTY);
+			nnInputs.add(0.0);
+			nnInputs.add(0.0);
+		}
+		return nnInputs;
+	}
+
+	protected boolean inSight(AbstractAgent agent) {
+		double crossProduct = this.cosTeta(this.getRx(), this.getRy(), agent.getX() - this.getX(), agent.getY() - this.getY());
+		return (crossProduct > 0);
+	}
+
+	protected double distanceTo(AbstractAgent agent) {
+		return this.module(agent.getX() - this.getX(), agent.getY() - this.getY());
+	}
+
+	protected double cosTeta(double vx1, double vy1, double vx2, double vy2) {
+		double v1 = this.module(vx1, vy1);
+		double v2 = this.module(vx2, vy2);
+		if (v1 == 0) {
+			v1 = 1E-5;
+		}
+		if (v2 == 0) {
+			v2 = 1E-5;
+		}
+		double ret = ((vx1 * vx2) + (vy1 * vy2)) / (v1 * v2);
+		return ret;
+	}
+
+	protected double module(double vx1, double vy1) {
+		return Math.sqrt((vx1 * vx1) + (vy1 * vy1));
+	}
+
+	protected double pseudoScalarProduct(double vx1, double vy1, double vx2, double vy2) {
+		return (vx1 * vy2) - (vy1 * vx2);
+	}
+
+	private double normalizeSpeed(double speed) {
+		double abs = Math.abs(speed);
+		if (abs > maxSpeed) {
+			double sign = Math.signum(speed);
+			speed = sign * (abs - (Math.floor(abs / maxSpeed) * maxSpeed));
+		}
+		return speed;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 215:220, 224:229.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.